src/workerd/api/container.c++

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2025 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "container.h"
+
+#include <workerd/api/http.h>
+#include <workerd/io/io-context.h>
+
+namespace workerd::api {
+
+// =======================================================================================
+// Basic lifecycle methods
+
+Container::Container(rpc::Container::Client rpcClient)
+    : rpcClient(IoContext::current().addObject(kj::heap(kj::mv(rpcClient)))) {}
+
+void Container::start(jsg::Lock& js, jsg::Optional<StartupOptions> maybeOptions) {
+  JSG_REQUIRE(!running, Error, "start() cannot be called on a container that is already running.");
+
+  StartupOptions options = kj::mv(maybeOptions).orDefault({});
+
+  auto req = rpcClient->startRequest();
+  KJ_IF_SOME(entrypoint, options.entrypoint) {
+    auto list = req.initEntrypoint(entrypoint.size());
+    for (auto i: kj::indices(entrypoint)) {
+      list.set(i, entrypoint[i]);
+    }
+  }
+  req.setEnableInternet(options.enableInternet);
+
+  IoContext::current().addTask(req.send().ignoreResult());
+
+  running = true;
+}
+
+jsg::Promise<void> Container::monitor(jsg::Lock& js) {
+  JSG_REQUIRE(running, Error, "monitor() cannot be called on a container that is not running.");
+
+  return IoContext::current()
+      .awaitIo(js, rpcClient->monitorRequest(capnp::MessageSize{4, 0}).send().ignoreResult())
+      .then(js, [this](jsg::Lock& js) {
+    running = false;
+    KJ_IF_SOME(d, destroyReason) {
+      jsg::Value error = kj::mv(d);
+      destroyReason = kj::none;
+      js.throwException(kj::mv(error));
+    }
+  }, [this](jsg::Lock& js, jsg::Value&& error) {
+    running = false;
+    destroyReason = kj::none;
+    js.throwException(kj::mv(error));
+  });
+}
+
+jsg::Promise<void> Container::destroy(jsg::Lock& js, jsg::Optional<jsg::Value> error) {
+  if (!running) return js.resolvedPromise();
+
+  if (destroyReason == kj::none) {
+    destroyReason = kj::mv(error);
+  }
+
+  return IoContext::current().awaitIo(
+      js, rpcClient->destroyRequest(capnp::MessageSize{4, 0}).send().ignoreResult());
+}
+
+void Container::signal(jsg::Lock& js, int signo) {
+  JSG_REQUIRE(signo > 0 && signo <= 64, TypeError, "Invalid signal number.");
```

Comment:
Nit: should be a `RangeError`?

Reply:
6fcd2f3f4a3f1d869194ebc8775c1fcee018bc5e